src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -1252,5 +1271,170 @@ internal bool CanAddParsedValue(HttpHeaderParser parser)
 
             internal bool IsEmpty => (RawValue == null) && (InvalidValue == null) && (ParsedValue == null);
         }
+
+        internal struct HeaderStore
+        {
+            private const int InitialCapacity = 4;
+            private const int MaxHeaderCount = 128;
+
+            public HeaderEntry[]? Entries;
```

Comment:
Might be crazy idea but:
How common is the `Count == 0` case in practice? If less common than `Count > 0`, isn't it worth to define an embedded storage to avoid the array allocation?

```C#
internal struct HeaderStore
{
    private HeaderEntry _e000, _e001, _e002, _e003, _e004, _e005 /*...*/ ;

    public Span<HeaderEntry> Entries => MemoryMarshal.CreateSpan<HeaderEntry>(ref _e0, MaxHeaderCount);
}
```

Reply:
We definitely wouldn't want to store the entire max as fields since having few headers is very common. But something like storing the first 6 headers as fields and deferring to an array when more are added is something we could potentially explore. I imagine it would come down to how much it would complicate the logic vs. saving the small allocation.